presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,114 @@
+
+            Map<Expression, FieldId> columnReferenceFields = analysis.getColumnReferenceFields();
+            List<Expression> columnReferences = ImmutableList.copyOf(analysis.getColumnReferences());
+            List<Expression> groupingOrdinals = expression.getGroupingColumns().stream()
+                    .map(groupingColumn -> extractGroupingColumnOrdinal(groupingColumn, columnReferences, columnReferenceFields))
+                    .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                    .collect(toImmutableList());
+
+            List<List<Expression>> groupingSetOrdinals;
+            ImmutableList.Builder<List<Expression>> groupingSetOrdinalsBuilder = ImmutableList.builder();
+            for (List<Expression> groupingSet : analysis.getGroupingSets(queryNode)) {
+                groupingSetOrdinalsBuilder.add(
+                        groupingSet.stream()
+                            .map(groupingColumn -> extractGroupingColumnOrdinal(groupingColumn, columnReferences, columnReferenceFields))
+                            .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                            .collect(toImmutableList()));
+            }
+            groupingSetOrdinals = groupingSetOrdinalsBuilder.build();
+
+            Expression firstGroupingArgument = groupIdSymbol.get().toSymbolReference();
+            List<Expression> newGroupingArguments = ImmutableList.of(
+                    firstGroupingArgument,
+                    new Cast(new ArrayConstructor(groupingOrdinals), ListLiteralType.NAME),
+                    new Cast(new ArrayConstructor(groupingSetOrdinals.stream().map(ArrayConstructor::new).collect(toImmutableList())), ListLiteralType.NAME)
+            );
+
+            FunctionCall rewritten = new FunctionCall(expression.getLocation().get(), QualifiedName.of(GROUPING), newGroupingArguments);
+            List<TypeSignatureProvider> functionArgumentTypes = Arrays.asList(
+                    new TypeSignatureProvider(BIGINT.getTypeSignature()),
+                    new TypeSignatureProvider(ListLiteralType.LIST_LITERAL.getTypeSignature()),
+                    new TypeSignatureProvider(ListLiteralType.LIST_LITERAL.getTypeSignature())
+            );
+            resolveFunction(rewritten, functionArgumentTypes, metadata.getFunctionRegistry());
+
+            return rewritten;
+        }
+    }
+
+    // TODO: This method is a hack and is needed until Analysis#columnReferences is converted to a IdentityLinkedHashMap.
+    // The first line fetches the columnReference instance using equality. That instance is then used to lookup the FieldId
+    // using an identity comparison.
+    private static int extractGroupingColumnOrdinal(Expression groupingExpression, List<Expression> columnReferences, Map<Expression, FieldId> columnReferenceFields)
+    {
+        Expression lookupExpression = columnReferences.get(columnReferences.indexOf(groupingExpression));
```

Comment:
`columnReferences.indexOf` compares by equality. Related discussion @ https://github.com/Teradata/presto/commit/5d8611dd284d74327ee7874dcb694d2d5b067989#commitcomment-21992584.

I'll copy my last comment from there so it doesn't get lost when GH garbagecollects the commit:

---

[Note] that `columnReferences` can contain equal (but not the same) `Identitfier` multiple times, referring to one or multiple different columns. For example in query

```
select a, a+1, 
  (select a from (values 37) u(a))
from (values 42) t(a)
```

we have 3 instances of `Identifier("a")`. Two of them refer to first column of `t` table (42) and one does not. Which one should be found?

If you have newly created instances of either `Identifier` or `DereferenceExpression`, perhaps you should analyze them, using proper `Scope`.

@sopel39 can you help? I think you did something like this?